### DISCUSSION TITLE
Extending without calling twice

### DISCUSSION THREAD

**Discussion by maucaro** [fingerprintjs/fingerprintjs#757] [Q&A]
In the extension [examples](https://github.com/fingerprintjs/fingerprintjs/blob/master/docs/extending.md), one needs to generate a fingerprint in order to get the components and then generate another fingerprint, in essence generating the fingerprint twice. Is there a way to only do so only once?

**Comment by Finesse** [comment]
Hi @maucaro. Generating a fingerprint consists of 2 steps:

1. Collecting the entropy components (the `components` field of the result object)
2. Calculating a hash sum of the components (the `visitorId` field of the result object)

The `visitorId` field is lazy. It means that the hash won't be calculated until you read the field.

So, if you follow the extension guide, the second step will be skipped, and the hash will be calculated only once. What will happen:

1. The built-in entropy components are collected
2. Your custom components are collected
3. A hash sum of all the components is calculated

No excess job will be done when you add custom components. Excess job will be done when you exclude a component because it is collected regardless. This is [a known issue](https://github.com/fingerprintjs/fingerprintjs/issues/718), we will handle it in future.

**Reply by maucaro** [reply]
@Finesse , thanks for the explanation. Collecting the entropy components can be expensive and I was looking into dropping some based on this cost (fonts for example). But if I have incurred the cost of collecting the full set, I won't save much, if anything by excluding it. Ideally I'd like to be able to express that I only want to collect a certain subset of entropy components and then generate the visitorId. If this can't be done today, so be it, not a huge deal. But, it would be nice :-)

**Reply by Finesse** [reply]
So, https://github.com/fingerprintjs/fingerprintjs/issues/718 very close to your needs. Please upvote it so that we know that it's important.

**Reply by maucaro** [reply]
Done. Thanks

**Reply by JsfasdF252** [reply]
#718 E Tigercat cutting machine would for sure !
